File: javascript/lesson 11/subsets.js
// https://leetcode.com/problems/subsets/

function subsets(nums) {
    const result = [];
    subsetsImpl(nums, 0, [], result);
    return result;
}

function subsetsImpl(nums, start, current, result) {
    result.push([...current]);
        
    for (let i = start; i < nums.length; i++) {
        current.push(nums[i]);
        subsetsImpl(nums, i + 1, current, result);
        current.pop();
    }
}

File: ruby/lesson 11/subsets.rb
# https://leetcode.com/problems/subsets/

def subsets(nums)
    result = []
    subsets_impl(nums, 0, [], result)
    result
end

def subsets_impl(nums, start, current, result)
    result.push(current.clone)
    
    (start..nums.length - 1).each do |i|
        current.push(nums[i])
        subsets_impl(nums, i + 1, current, result)
        current.pop()
    end
end
